spec/java/util/HashMap.main.lsl

```diff
@@ -0,0 +1,642 @@
+            val expectedModCount: int = this.modCount;
+            newValue= action CALL(remappingFunction, [oldValue, newValue]);
+            _checkForComodification(expectedModCount);
+            if (newValue == null)
+            {
+                result = action MAP_GET(this.storage, key);
+                action MAP_REMOVE(this.storage, key);
+            }
+            else
+            {
+                action CALL_METHOD(entry, "setValue", [newValue]);
+            }
+
+            this.modCount += 1;
+            result = newValue;
+        }
+        else
+        {
+            action CALL_METHOD(entry, "setValue", [newValue]);
+            this.modCount += 1;
+            result = value;
+        }
+    }
+
+
+    fun *.put (@target self: HashMap, key: Object, value: Object): Object
+    {
+        result = null;
+        if (action MAP_HAS_KEY(this.storage, key))
+        {
+            var entry: Map_Entry<Object, Object> = action MAP_GET(this.storage, key);
+            result = action CALL_METHOD(entry, "getValue", []);
+            action CALL_METHOD(entry, "setValue", [value]);
+        }
+        else
+        {
+            val newEntry: Map_Entry<Object, Object> = new SimpleEntryAutomaton(state = Initialized,
+                key = key,
+                value = value
+            );
+            action MAP_SET(this.storage, key, newEntry);
+        }
+        this.modCount += 1;
+    }
+
+
+    fun *.putAll (@target self: HashMap, m: Map): void
+    {
+        if (m == null)
+            _throwNPE();
+
+        val mSize: int = action CALL_METHOD(m, "size", []);
+        if (mSize != 0)
+            _addAllElements(m);
+    }
+
+
+    fun *.putIfAbsent (@target self: HashMap, key: Object, value: Object): Object
+    {
+        result = _getMappingOrDefault(key, null);
+        if (result == null && action MAP_HAS_KEY(this.storage, key))
+        {
+            val entry: Map_Entry<Object, Object> = action MAP_GET(this.storage, key);
+            action CALL_METHOD(entry, "setValue", [value]);
+            this.modCount += 1;
+        }
+    }
+
+
+    fun *.remove (@target self: HashMap, key: Object): Object
+    {
+        if (action MAP_HAS_KEY(this.storage, key))
+        {
+            val entry: Map_Entry<Object, Object> = action MAP_GET(this.storage, key);
+            result = action CALL_METHOD(entry, "getValue", [])
+            action MAP_REMOVE(this.storage, key);
+            this.modCount += 1;
+        }
+        else
+        {
+            result = null;
+        }
+    }
+
+
+    fun *.remove (@target self: HashMap, key: Object, value: Object): boolean
+    {
+        result = false;
+
+        if (action MAP_HAS_KEY(this.storage, key))
+        {
+            val entry: Map_Entry<Object, Object> = action MAP_GET(this.storage, key);
+            val curValue: Object = action CALL_METHOD(entry, "getValue", []);
+            // #question: It will throw NPE if curValue == null and value == null ? Or not ?
```

Comment:
[No](https://github.com/UnitTestBot/java-stdlib-approximations/blob/main/approximations/src/main/java/runtime/LibSLRuntime.java#L894), it will not.

Reply:
Comment was deleted (fixed)